pkgs/shells/fish/default.nix

```diff
@@ -35,28 +35,60 @@ let
     #     source both, but source the more global configuration files earlier
     #     than the more local ones, so that more local configurations inherit
     #     from but override the more global locations.
-    
+
     if test -f /etc/fish/config.fish
       source /etc/fish/config.fish
     end
-    
+
     #                                                                             #
     ############### ↑ Nix hook for sourcing /etc/fish/config.fish ↑ ###############
   '';
 
+
+  nixEnvironmentSetupScriptPrefix = "@NIX_STORE@/../var/nix/profiles/default/etc/profile.d";
+  selectPosixPreInitScript = if hostPlatform.isDarwin then ''
+    # anchor for indentation
+        # Use `set -U nix_on_darwin_no_daemon` to tell Fish to use the single-user mode script instead
+        if test -f ${nixEnvironmentSetupScriptPrefix}/nix-daemon.sh; and not set -q nix_on_darwin_no_daemon
+          fenv source ${nixEnvironmentSetupScriptPrefix}/nix-daemon.sh
+        else
+          fenv source ${nixEnvironmentSetupScriptPrefix}/nix.sh
+        end
+  '' else ''
+    # anchor for indentation
+        test -f ${nixEnvironmentSetupScriptPrefix}/nix.sh
+        and fenv source ${nixEnvironmentSetupScriptPrefix}/nix.sh
+  '';
   fishPreInitHooks = ''
-    # source nixos environment
+    # source nix environment
     # note that this is required:
     #   1. For all shells, not just login shells (mosh needs this as do some other command-line utilities)
     #   2. Before the shell is initialized, so that config snippets can find the commands they use on the PATH
-    builtin status --is-login
-    or test -z "$__fish_nixos_env_preinit_sourced" -a -z "$ETC_PROFILE_SOURCED" -a -z "$ETC_ZSHENV_SOURCED"
-    and test -f /etc/fish/nixos-env-preinit.fish
-    and source /etc/fish/nixos-env-preinit.fish
-    and set -gx __fish_nixos_env_preinit_sourced 1
-
-    test -n "$NIX_PROFILES"
-    and begin
+    if builtin status --is-login
+       or test -z "$__fish_nix_env_preinit_sourced" -a -z "$ETC_PROFILE_SOURCED" -a -z "$ETC_ZSHENV_SOURCED"
```

Comment:
This duplication is maybe worth fixing now, maybe something to keep an eye on for later. 🤷‍♂️

Reply:
I think this should be fixed, but it will mean changing things for `bash`, `zsh`, etc., in NixOS (and maybe Nix, for its scripts). The change could be as trivial as settling on a single variable in all of them, but it still seems out of scope for this PR.

Reply:
Maybe the various shells should have a variable in common they use for this. But that's an issue that involves other shells and other parts of Nixpkgs. I think we can set it aside.

When these lines were essentially part of the NixOS module, I checked these variables for the same reason that the NixOS module does: we don't want to clobber configuration made elsewhere in NixOS or redundantly source things that have already been sourced from other shells.

But here, it's only used for the basic `set-environment.sh` script, which doesn't really include any user configuration. And on non-NixOS, those variables might never be set anyway. Here, maybe it makes more sense to just check NIX_PROFILES.

Reply:
Yup yup. I've logged a separate issue about cleaning this up. #31322